Feature/FARM-381 playground chat integration

alexanishchenko88 opened the pull request:
[FARM-381 - Integrate chat]

📲 What
Chat playground has been created.

🤔 Why
Use should be able to interact with chat bot. This bot will provide a couple of question what user should answer by the way to improve the experience.

🛠 How
Chat has been created as a separated folder. Everything related to chat is under lib/chat. This way in the future it going to be an isolated SDK.
Chat works with provided File right now. Also have callbacks to collect the success or error.
In the playground we can see the chat under Playground/SPIKE Chat Page.

Chat features:

Will start with header message once the view has been rendered.
Will show the welcome message and then a sequence of interactive messages.
User will answer the questions by typing the answer or taping on one of the provided options.
Once finished, it will show the summary view with answers.
Once reviewed the answers and pressed the confirmation button it will close the chat view and via callbacks will notify the provided answers.

#### 👀 See

<img width="50%" alt="Premium User" src="https://user-images.githubusercontent.com/40355600/62374947-64c86d00-b53d-11e9-8f49-466108495145.gif">

<img width="50%" alt="Premium User" src="https://user-images.githubusercontent.com/40355600/62374940-5e39f580-b53d-11e9-982e-1a33b27079ed.gif">

		 
#### ✅ Acceptance criteria

- [ ] Design Review for UI with BA completed. 
- [ ] Manually tested and verified on Android device/emulator (min. Lollipop 5.1)
- [ ] Showcase video / automation test
- [ ] Passing all tests
- [ ] Rebased/merged with latest changes from development and re-tested
- [ ] Removed unsed comments. TODOs must have JIRA for future work.

#### Coding Standards Checklist
- [ ] unit tests 80% coverage on testable code (functions, methods, class)
- [ ] (Architecture) redux state, reducers, actions, middleware defined under lib/redux
- [ ] ui elements defined under lib/ui. 
- [ ] Strings ready for localisation (e.g. defined in lib/utils/strings.dart)
- [ ] Assets (images/icons path) defined in lib/utils/assets.dart
- [ ] Colors defined in lib/utils/colors.dart
- [ ] Dimensions (ie margins padding) defined in lib/utils/dimens.dart
- [ ] TextStyles defined in lib/utils/styles.dart

Recommended Style Guide: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo

#### 🕵️‍♂️ How to test

Notes for QA

#### Reviewers

@farmsmart/amido @farmsmart/wamf

alexanishchenko88 committed:
[FARM-Chat] Playground for chat integration

Integration for chat. Using ChatPage to start chat page

alexanishchenko88 committed:
[FARM-Chat] Playground for chat integration

Wrongly removed .ttf files are back

davidfranquet commented on lib/chat/ChatPage.dart:
```diff
@@ -0,0 +1,34 @@
+  }
+
+  File _createChatFile() {
+    return File("assets/responses/create_account_mock_form.json");
```
move to private _Constants class or pass this param on ChatPage constructor

alexanishchenko88 replied:
Sure, will move it to private

alexanishchenko88 replied:
Done

davidfranquet commented on lib/chat/ChatPage.dart:
```diff
@@ -0,0 +1,34 @@
+
+  _doOnSuccess(BuildContext context, Map<String, String> map) {
+    print("On Success recevied: ${map.toString()}");
+    Navigator.pop(context);
```
```suggestion
    Navigator.of(context).pop();
```

alexanishchenko88 replied:
Is any difference on do it this way?

alexanishchenko88 replied:
In some parts of app I've seen it this way Navigator.pop(context) as for example in ActionSheet.
Will change it anyway.

davidfranquet commented on lib/chat/ChatPage.dart:
```diff
@@ -0,0 +1,34 @@
+
+  _doOnError(BuildContext context, String error) {
+    print("On Error recevied: ${error.toString()}");
+    Navigator.pop(context);
```
```suggestion
    Navigator.of(context).pop();
```

alexanishchenko88 replied:
Changed

davidfranquet commented on lib/chat/bloc/handler/implementation/InteractiveMessageHandlerImpl.dart:
```diff
@@ -0,0 +1,161 @@
+    bool isFocusedOnBuild,
+  }) {
+    return TextInput(
+      buttonText: _Strings.sendText,
```
Should be localized
```suggestion
      buttonText: Intl.message(_Strings.sendText),
```

alexanishchenko88 replied:
Changed in whole file

davidfranquet commented on lib/chat/bloc/handler/implementation/InteractiveMessageHandlerImpl.dart:
```diff
@@ -0,0 +1,161 @@
+  }
+
+  InputDecoration _getInputDecoration() {
+    return InputDecoration(hintText: _Strings.inputHint);
```
```suggestion
    return InputDecoration(hintText: Intl.message(_Strings.inputHint));
```

alexanishchenko88 replied:
Changed in whole file

davidfranquet commented on lib/chat/bloc/handler/implementation/InteractiveMessageHandlerImpl.dart:
```diff
@@ -0,0 +1,161 @@
+    Function(String) onValidationPassed,
+  }) {
+    if (value.isEmpty) {
+      return _Strings.messageNotEmpty;
```
Check all the strings and localise them: `Intl.message(string)`

alexanishchenko88 replied:
Changed in whole file

davidfranquet commented on lib/chat/bloc/helper/implementation/ChatMessageProviderHelperImpl.dart:
```diff
@@ -0,0 +1,72 @@
+class ChatMessageProviderHelperImpl
+    implements ChatMessageProviderHelper<MessageBubbleViewModel> {
+  @override
+  MessageBubbleViewModel getMessageFromEntity(FormItemEntity entity) {
```
It's ok, but remember that you can do this: 
```suggestion
  MessageBubbleViewModel getMessageFromEntity(FormItemEntity entity) => _getMessageAsBubble(entity);
```

alexanishchenko88 replied:
Maybe we should have some way of make it as a standard. For me is much clear to read with curly brackets {} instead of calling **=>**

Let's decide this and will change to any of 2 options (or maybe some third one 😄 )

LeeMatthewHiggins replied:
yeah I have the same issue, the brackets are clearer for me, but seems the standard wants the shorthand.

alexanishchenko88 replied:
Ok, in this case I will change it to the standard way in case we have small logic as in this case could be a function.

davidfranquet commented on lib/chat/bloc/helper/implementation/ChatSummaryProviderHelperImpl.dart:
```diff
@@ -0,0 +1,35 @@
+  String _getInputModelAsString({Map<String, String> inputModel}) {
+    String response = "";
+    inputModel.forEach((key, value) {
+      response = response + key + " : " + value + "\n";
```
```suggestion
      response = '$response$key : $value\n';
```

alexanishchenko88 replied:
Changed

davidfranquet commented on lib/chat/bloc/provider/ChatProvider.dart:
```diff
@@ -0,0 +1,307 @@
+  static const String summaryError = "Provided summary is not correct";
+}
+
+class ChatProvider implements ViewModelProvider<ChatViewModel> {
```
Check strings, should be localised

alexanishchenko88 replied:
Changed in whole file

davidfranquet commented on lib/chat/ui/widgets/bubble_message.dart:
```diff
@@ -0,0 +1,184 @@
+    MessageBubbleStyle style,
+    Function onTap,
+  })  : this._viewModel = viewModel,
+        this._style = style ?? _defaultStyle,
```
```suggestion
        this._style = style,
```

alexanishchenko88 replied:
Hmm, no default style for widget? Any reason of do it this way?

alexanishchenko88 replied:
Just seen the message bellow. Will change it.

davidfranquet commented on lib/chat/ui/widgets/bubble_message.dart:
```diff
@@ -0,0 +1,184 @@
+
+  MessageBubble({
+    @required MessageBubbleViewModel viewModel,
+    MessageBubbleStyle style,
```
Add the defaultStyle here instead of adding below. You won't need to check if the incoming style is null or not
```suggestion
    MessageBubbleStyle style = _defaultStyle,
```

davidfranquet commented on lib/chat/ui/widgets/bubble_message.dart:
```diff
@@ -0,0 +1,184 @@
+            children: <Widget>[
+              _buildAvatar(),
+              _buildContainer(),
+            ].where(_notNull).toList(),
```
👍

davidfranquet commented on lib/chat/ui/widgets/bubble_message.dart:
```diff
@@ -0,0 +1,184 @@
+        ));
+  }
+
+  bool _notNull(Widget item) => item != null;
```
move this to the top (below the finals)

alexanishchenko88 replied:
Moved. Do not really understand the purpose of doing so, is a function...

davidfranquet commented on lib/chat/ui/widgets/card_view.dart:
```diff
@@ -0,0 +1,91 @@
+  final CardViewStyle _style;
+  final Widget _child;
+
+  CardView({CardViewStyle style, Widget child})
```
Add the defaultStyle here instead of adding below. You won't need to check if the incoming style is null or not

alexanishchenko88 replied:
Changed

davidfranquet commented on lib/chat/ui/widgets/chat.dart:
```diff
@@ -0,0 +1,97 @@
+
+  Chat({
+    @required ChatProvider chatProvider,
+    ChatStyle style,
```
Add the defaultStyle here instead of adding below. You won't need to check if the incoming style is null or not

alexanishchenko88 replied:
Changed

davidfranquet commented on lib/chat/ui/widgets/selectable_options.dart:
```diff
@@ -0,0 +1,143 @@
+  Widget build(BuildContext context) {
+    return Container(
+      margin: _style.outerContainerMargin,
+      child: Wrap(
```
👍

davidfranquet commented on lib/chat/utils/delay_tween.dart:
```diff
@@ -0,0 +1,22 @@
+import 'dart:math' as math show sin, pi;
+import 'package:flutter/widgets.dart';
+
+class DelayTween extends Tween<double> {
```
Add the unit test. (It's already done)

alexanishchenko88 replied:
Added

alexanishchenko88 committed:
[FARM-Chat] Playground for chat integration

Unit tests have been added for 3 dots iterator and for some JSON data serialisable mappers.

albertmiro commented on lib/chat/bloc/provider/ChatProvider.dart:
```diff
@@ -0,0 +1,307 @@
+          _chatViewModel.interactiveWidget =
+              _buildSelectableOptionsWidget(entity: entity);
+          break;
+      }
```
I think that this should be the same:

```
      InteractiveMessageType inputType;
      switch (entity.type) {
        case _Constants.typeValueString:
          inputType = InteractiveMessageType.inputString;
          break;
        case _Constants.typeValueEmail:
          inputType = InteractiveMessageType.inputEmail;
          break;
        case _Constants.typeValuePhoneNumber:
          inputType = InteractiveMessageType.inputPhoneNumber;
          break;
        case _Constants.typeValueImage:
          inputType = InteractiveMessageType.inputImage;
          break;
        case _Constants.typeValueMultiChoice:
          inputType = null;
          break;
      }

      _chatViewModel.interactiveWidget = _buildInputTextWidget(
        entity: entity,
        inputType: inputType,
      );
```

albertmiro replied:
It would be nice if we can have enums with values, so then we would reduce this switch, not sure if we can do it with flutter.

alexanishchenko88 replied:
Refactored into 2 functions to make it more readable.
Do not see the need of having the value for enum, I just need to know the type of provided input. Could you provide some additional info?

LeeMatthewHiggins replied:
👍

alexanishchenko88 committed:
[FARM-Chat] Playground for chat integration

PR comments resolved.

Some minor changes on code structure.
Intl localisation added.

alexanishchenko88 committed:
[FARM-Chat] Playground for chat integration

PR comments.
Some refactor on ChatProvider class

LeeMatthewHiggins commented on lib/chat/bloc/ViewModelProvider.dart:
```diff
@@ -0,0 +1,9 @@
+import 'dart:async';
+
+abstract class ViewModelProvider<T> {
```
this exists in the project already, how do we deal with that I wonder? Is the chat product in its own scope?

alexanishchenko88 replied:
Having Chat as something isolated (as SDK) maybe we should consider having this in two places. Otherwise we should add a dependency o start thinking in some new workaround for that. 
What is you opinion guys? @albertmiro @DavidFD005 @LeeMatthewHiggins

albertmiro replied:
That is not generating name conflicts, we would leave that like it is for now.

LeeMatthewHiggins reviewed (dismissed):
Lets get this in, check the conflicting names for the classes first.

albertmiro reviewed (dismissed):
Design would be done in a separate PR.

alexanishchenko88 committed:
[FARM-Chat] Playground for chat integration

Changing some logic in returning on functions in case functions have a simple logic.

LeeMatthewHiggins reviewed (dismissed):
LGTM

Commit:
[FARM-381] Trigger CI build

davidfranquet reviewed (approved):
🚀
